File: app/src/main/java/br/com/valenstech/letraviva/util/ValidacaoConteudo.java
package br.com.valenstech.letraviva.util;

import android.webkit.URLUtil;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.regex.Pattern;

public final class ValidacaoConteudo {

    private static final Pattern PADRAO_TAG_HTML = Pattern.compile("<[^>]+>", Pattern.CASE_INSENSITIVE);
    private static final Pattern PADRAO_SCRIPT_OU_EVENTO = Pattern.compile(
            "(?i)(<\\s*/?\\s*script|&lt;\\s*/?\\s*script|&#x3c;\\s*/?\\s*script|javascript:|data:text/html|on[a-z]+\\s*=)"
    );
    private static final Pattern PADRAO_CARACTERE_CONTROLE = Pattern.compile("[\\p{C}&&[^\\n\\r\\t]]");
    private static final Pattern PADRAO_URL_HTTP_HTTPS = Pattern.compile("^https?://.+", Pattern.CASE_INSENSITIVE);

    private ValidacaoConteudo() {
        // Utilitário
    }

    public static boolean possuiHtmlSuspeito(@NonNull String texto) {
        String analisado = texto.trim();
        return PADRAO_TAG_HTML.matcher(analisado).find() || PADRAO_SCRIPT_OU_EVENTO.matcher(analisado).find();
    }

    public static boolean validarTextoSeguroObrigatorio(@Nullable String texto) {
        if (texto == null || texto.trim().isEmpty()) {
            return false;
        }
        return !possuiHtmlSuspeito(texto);
    }

    public static boolean validarTextoSeguroOpcional(@Nullable String texto) {
        if (texto == null || texto.trim().isEmpty()) {
            return true;
        }
        return !possuiHtmlSuspeito(texto);
    }

    public static boolean validarUrlSegura(@Nullable String url) {
        if (url == null || url.trim().isEmpty()) {
            return false;
        }
        String limpa = sanitizarBasico(url);
        if (!PADRAO_URL_HTTP_HTTPS.matcher(limpa).matches()) {
            return false;
        }
        if (!URLUtil.isValidUrl(limpa)) {
            return false;
        }
        return !possuiHtmlSuspeito(limpa);
    }

    @NonNull
    public static String sanitizarBasico(@NonNull String texto) {
        return PADRAO_CARACTERE_CONTROLE.matcher(texto).replaceAll("").trim();
    }

    @NonNull
    public static String aplicarEscapeSeguro(@NonNull String texto) {
        if (texto.isEmpty()) {
            return "";
        }
        StringBuilder resultado = new StringBuilder(texto.length());
        for (int indice = 0; indice < texto.length(); indice++) {
            char caractere = texto.charAt(indice);
            switch (caractere) {
                case '<':
                    resultado.append('\uFF1C');
                    break;
                case '>':
                    resultado.append('\uFF1E');
                    break;
                case '&':
                    resultado.append('\uFF06');
                    break;
                case '\"':
                    resultado.append('\uFF02');
                    break;
                case '\'':
                    resultado.append('\uFF07');
                    break;
                default:
                    resultado.append(caractere);
                    break;
            }
        }
        return resultado.toString();
    }
}

File: app/src/main/java/br/com/valenstech/letraviva/util/ValidacaoConteudo.kt
package br.com.valenstech.letraviva.util

import android.webkit.URLUtil

object ValidacaoConteudo {

    private val padraoTagHtml = Regex("<[^>]+>", RegexOption.IGNORE_CASE)
    private val padraoScriptOuEvento = Regex("(?i)(<\\s*/?\\s*script|&lt;\\s*/?\\s*script|&#x3c;\\s*/?\\s*script|javascript:|data:text/html|on[a-z]+\\s*=)")
    private val padraoCaracterControle = Regex("[\\p{C}&&[^\\n\\r\\t]]")
    private val padraoUrlHttpHttps = Regex("^https?://.+", RegexOption.IGNORE_CASE)

    fun possuiHtmlSuspeito(texto: String): Boolean {
        val analisado = texto.trim()
        return padraoTagHtml.containsMatchIn(analisado) || padraoScriptOuEvento.containsMatchIn(analisado)
    }

    fun validarTextoSeguroObrigatorio(texto: String?): Boolean {
        if (texto.isNullOrBlank()) {
            return false
        }
        return !possuiHtmlSuspeito(texto)
    }

    fun validarTextoSeguroOpcional(texto: String?): Boolean {
        if (texto.isNullOrBlank()) {
            return true
        }
        return !possuiHtmlSuspeito(texto)
    }

    fun validarUrlSegura(url: String?): Boolean {
        if (url.isNullOrBlank()) {
            return false
        }
        val limpa = sanitizarBasico(url)
        if (!padraoUrlHttpHttps.matches(limpa)) {
            return false
        }
        if (!URLUtil.isValidUrl(limpa)) {
            return false
        }
        if (possuiHtmlSuspeito(limpa)) {
            return false
        }
        return true
    }

    fun sanitizarBasico(texto: String): String {
        return texto.replace(padraoCaracterControle, "").trim()
    }

    fun aplicarEscapeSeguro(texto: String): String {
        if (texto.isEmpty()) {
            return ""
        }
        val resultado = StringBuilder(texto.length)
        texto.forEach { caractere ->
            when (caractere) {
                '<' -> resultado.append('\uFF1C')
                '>' -> resultado.append('\uFF1E')
                '&' -> resultado.append('\uFF06')
                '\"' -> resultado.append('\uFF02')
                '\'' -> resultado.append('\uFF07')
                else -> resultado.append(caractere)
            }
        }
        return resultado.toString()
    }
}
